lang/funcs/collection.go

```diff
@@ -106,6 +107,145 @@ var LengthFunc = function.New(&function.Spec{
 	},
 })
 
+// CoalesceListFunc contructs a function that takes any number of list arguments
+// and returns the first one that isn't empty.
+var CoalesceListFunc = function.New(&function.Spec{
+	Params: []function.Parameter{},
+	VarParam: &function.Parameter{
+		Name:             "vals",
+		Type:             cty.DynamicPseudoType,
```

Comment:
We'd said we'd make this one a `cty.List(cty.DynamicPseudoType)`, right?

While in general we like to support both tuples and lists equivalently, in this case I think requiring a list is best because it then allows us to properly infer the return type of the function using the list element types.

In particular, unifying `cty.Tuple(["a", "b"])` with `cty.Tuple([])` will fail because the length of a tuple value is part of its type, and there's no reason to use this function if you know your lists are all the same length, because then either _all_ or _none_ of them will be empty.